Add a common base class for the rect-based BasicShapes

smfr opened the pull request:
#### c00e5b6fc5b38d5c215fd00a555b51c486010439
<pre>
Add a common base class for the rect-based BasicShapes
<a href="https://bugs.webkit.org/show_bug.cgi?id=282289">https://bugs.webkit.org/show_bug.cgi?id=282289</a>
<a href="https://rdar.apple.com/138872209">rdar://138872209</a>

Reviewed by Tim Nguyen.

There&apos;s data and code duplication between BasicShapeRect, BasicShapeXywh, and BasicShapeInset,
so add BasicRoundedRectShape and have those classes derive from it. This class can now handle
generation of rounded rects and shapes.

* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicRoundedRectShape::BasicRoundedRectShape):
(WebCore::BasicRoundedRectShape::operator== const):
(WebCore::BasicRoundedRectShape::roundedRectForBoundingBox const):
(WebCore::BasicRoundedRectShape::path const):
(WebCore::BasicShapeInset::create):
(WebCore::BasicShapeInset::BasicShapeInset):
(WebCore::BasicShapeInset::clone const):
(WebCore::BasicShapeInset::operator== const):
(WebCore::BasicShapeInset::rectForBoundingBox const):
(WebCore::BasicShapeInset::canBlend const):
(WebCore::BasicShapeInset::blend const):
(WebCore::BasicShapeInset::dump const):
(WebCore::operator&lt;&lt;):
(WebCore::BasicShapeRect::BasicShapeRect):
(WebCore::BasicShapeRect::operator== const):
(WebCore::BasicShapeRect::rectForBoundingBox const):
(WebCore::BasicShapeXywh::BasicShapeXywh):
(WebCore::BasicShapeXywh::operator== const):
(WebCore::BasicShapeXywh::rectForBoundingBox const):
(WebCore::m_bottomLeftRadius): Deleted.
(WebCore::BasicShapeRect::path const): Deleted.
(WebCore::BasicShapeXywh::path const): Deleted.
(WebCore::BasicShapeInset::path const): Deleted.
* Source/WebCore/rendering/style/BasicShapes.h:
(WebCore::BasicShape::isCircleTypeShape const):
(WebCore::BasicShape::isRectTypeShape const):
(WebCore::BasicRoundedRectShape::topLeftRadius const):
(WebCore::BasicRoundedRectShape::topRightRadius const):
(WebCore::BasicRoundedRectShape::bottomRightRadius const):
(WebCore::BasicRoundedRectShape::bottomLeftRadius const):
(WebCore::BasicRoundedRectShape::setTopLeftRadius):
(WebCore::BasicRoundedRectShape::setTopRightRadius):
(WebCore::BasicRoundedRectShape::setBottomRightRadius):
(WebCore::BasicRoundedRectShape::setBottomLeftRadius):

Canonical link: <a href="https://commits.webkit.org/285876@main">https://commits.webkit.org/285876@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/337be9646bced53eacb9f90b93569ea1dc71ae93

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74092 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26903 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78450 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25330 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76209 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62654 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1306 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16598 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63735 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38648 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45275 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21231 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23663 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66781 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21577 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79984 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1409 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/769 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1553 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63753 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16314 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9753 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7921 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1373 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1402 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1390 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1409 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -81,6 +82,18 @@ class BasicShape : public RefCounted<BasicShape> {
 
     virtual Type type() const = 0;
 
+    bool isCircleTypeShape() const
```
isEllipseTypeShape? technically circles are a type of ellipse, not the other way round

nt1m commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -408,46 +435,30 @@ class BasicShapeRect final : public BasicShape {
+    Type type() const override { return Type::Rect; }
+    FloatRect rectForBoundingBox(const FloatRect&) const override;
 
     bool canBlend(const BasicShape&) const final;
-    Ref<BasicShape> blend(const BasicShape& from, const BlendingContext&) const final;
+    Ref<BasicShape> blend(const BasicShape& from, const BlendingContext&) const override;
 
-    bool operator==(const BasicShape&) const final;
+    bool operator==(const BasicShape&) const override;
 
-    void dump(TextStream&) const final;
+    void dump(TextStream&) const override;
```
Aren't this part of the final class? Not sure why we need to change them to override. Same question for the other final classes

smfr replied:
`override` and `final` are equivalent if the class itself is marked `final`, so I made them consistent.

smfr committed:
Add a common base class for the rect-based BasicShapes
https://bugs.webkit.org/show_bug.cgi?id=282289
rdar://138872209

Reviewed by Tim Nguyen.

There's data and code duplication between BasicShapeRect, BasicShapeXywh, and BasicShapeInset,
so add BasicRoundedRectShape and have those classes derive from it. This class can now handle
generation of rounded rects and shapes.

* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicRoundedRectShape::BasicRoundedRectShape):
(WebCore::BasicRoundedRectShape::operator== const):
(WebCore::BasicRoundedRectShape::roundedRectForBoundingBox const):
(WebCore::BasicRoundedRectShape::path const):
(WebCore::BasicShapeInset::create):
(WebCore::BasicShapeInset::BasicShapeInset):
(WebCore::BasicShapeInset::clone const):
(WebCore::BasicShapeInset::operator== const):
(WebCore::BasicShapeInset::rectForBoundingBox const):
(WebCore::BasicShapeInset::canBlend const):
(WebCore::BasicShapeInset::blend const):
(WebCore::BasicShapeInset::dump const):
(WebCore::operator<<):
(WebCore::BasicShapeRect::BasicShapeRect):
(WebCore::BasicShapeRect::operator== const):
(WebCore::BasicShapeRect::rectForBoundingBox const):
(WebCore::BasicShapeXywh::BasicShapeXywh):
(WebCore::BasicShapeXywh::operator== const):
(WebCore::BasicShapeXywh::rectForBoundingBox const):
(WebCore::m_bottomLeftRadius): Deleted.
(WebCore::BasicShapeRect::path const): Deleted.
(WebCore::BasicShapeXywh::path const): Deleted.
(WebCore::BasicShapeInset::path const): Deleted.
* Source/WebCore/rendering/style/BasicShapes.h:
(WebCore::BasicShape::isCircleTypeShape const):
(WebCore::BasicShape::isRectTypeShape const):
(WebCore::BasicRoundedRectShape::topLeftRadius const):
(WebCore::BasicRoundedRectShape::topRightRadius const):
(WebCore::BasicRoundedRectShape::bottomRightRadius const):
(WebCore::BasicRoundedRectShape::bottomLeftRadius const):
(WebCore::BasicRoundedRectShape::setTopLeftRadius):
(WebCore::BasicRoundedRectShape::setTopRightRadius):
(WebCore::BasicRoundedRectShape::setBottomRightRadius):
(WebCore::BasicRoundedRectShape::setBottomLeftRadius):

Canonical link: https://commits.webkit.org/285876@main
